pkg/util/startup/retry.go

```diff
@@ -0,0 +1,190 @@
+//
+// If your test fails because startup assertion failed, then there's a new code
+// path run during server startup which is not correctly wrapped in retry loop
+// and it could cause node startup failures. You need to identify what is called
+// from Server.PreStart and add retry accordingly.
+package startup
+
+import (
+	"context"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+// startupRetryKey context key used by assertions to track that code is running
+// inside startup retry loop.
+type startupRetryKey struct{}
+
+// startupRetryOpts is retry config for startup operations. When changing values
+// ensure that retriesBeforeLogging is updated accordingly as to not spam logs
+// before at least couple of seconds have passed.
+var startupRetryOpts = retry.Options{
+	InitialBackoff: 100 * time.Millisecond,
+	MaxBackoff:     3 * time.Second,
+	Multiplier:     2,
+}
+
+// retriesBeforeLogging is number of retries startup loop does before it starts
+// logging errors.
+const retriesBeforeLogging = 6
+
+// runningStartup is a counter showing number of concurrently running server
+// startups. If it is zero, then startups are finished and we should not try
+// to run startup retry checks.
+// This counter only makes sense and used when running TestCluster and is an
+// optimization to stop doing any potentially expensive assertions once servers
+// are running.
+// Mind that this is a best effort tracking and assertions themselves should
+// not give false positives/negatives regardless of this counter value.
+var runningStartup atomic.Int32
+
+// WithoutChecks adds startup tag to context to allow kv and sql executor access
+// from startup path without wrapping methods in RunWithRetry*.
+func WithoutChecks(ctx context.Context) context.Context {
+	return context.WithValue(ctx, startupRetryKey{}, "bypass retry check")
+}
+
+// RunWithRetry synchronously runs a function with retry of circuit
+// breaker errors on startup.
+func RunWithRetry(ctx context.Context, opName string, f func(ctx context.Context) error) error {
+	_, err := RunWithRetryEx(ctx, opName, func(ctx context.Context) (any, error) {
+		return nil, f(ctx)
+	})
+	return err
+}
+
+// RunWithRetryEx run function returning value with startup retry.
+func RunWithRetryEx[T any](
+	ctx context.Context, opName string, f func(ctx context.Context) (T, error),
+) (T, error) {
+	ctx = context.WithValue(ctx, startupRetryKey{}, "in retry")
+	// Retry indefinitely until context is cancelled.
+	var result T
+	var err error
+	for r := retry.Start(startupRetryOpts); r.Next(); {
+		result, err = f(ctx)
+		if err == nil {
+			break
+		}
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		if r.CurrentAttempt() > retriesBeforeLogging {
```

Comment:
This is ever so slightly clumsy, how about this:

```go
// at the beginning of this method
every := log.Every(5*time.Second)

for ... {
  // ...
  if every.ShouldLog() {
    log.Infof(...)
  }
}
```

It makes sense to log on the first retry, these retries should be rare and if they exist and become subject of investigation one would like to know the first time they occurred.

This approach also avoids the interdependence of `retriesBeforeLogging` and the retry opts.